rootston/output.c

```diff
@@ -25,293 +25,211 @@
  */
 void rotate_child_position(double *sx, double *sy, double sw, double sh,
 		double pw, double ph, float rotation) {
-	if (rotation != 0.0) {
-		// Coordinates relative to the center of the subsurface
-		double ox = *sx - pw/2 + sw/2,
-			oy = *sy - ph/2 + sh/2;
-		// Rotated coordinates
-		double rx = cos(rotation)*ox - sin(rotation)*oy,
-			ry = cos(rotation)*oy + sin(rotation)*ox;
-		*sx = rx + pw/2 - sw/2;
-		*sy = ry + ph/2 - sh/2;
+	if (rotation == 0.0) {
+		return;
 	}
+
+	// Coordinates relative to the center of the subsurface
+	double cx = *sx - pw/2 + sw/2,
+		cy = *sy - ph/2 + sh/2;
+	// Rotated coordinates
+	double rx = cos(rotation)*cx - sin(rotation)*cy,
+		ry = cos(rotation)*cy + sin(rotation)*cx;
+	*sx = rx + pw/2 - sw/2;
+	*sy = ry + ph/2 - sh/2;
 }
 
-struct layout_data {
-	double x, y;
+struct surface_iterator_data {
+	roots_surface_iterator_func_t user_iterator;
+	void *user_data;
+
+	struct roots_output *output;
+	double ox, oy;
 	int width, height;
 	float rotation;
 };
 
-static void get_layout_position(struct layout_data *data,
-		double *lx, double *ly, const struct wlr_surface *surface,
-		int sx, int sy) {
-	double _sx = sx, _sy = sy;
-	rotate_child_position(&_sx, &_sy, surface->current.width,
-		surface->current.height, data->width, data->height, data->rotation);
-	*lx = data->x + _sx;
-	*ly = data->y + _sy;
-}
+static bool get_surface_box(struct surface_iterator_data *data,
+		struct wlr_surface *surface, int sx, int sy,
+		struct wlr_box *surface_box) {
+	struct roots_output *output = data->output;
 
-static void surface_for_each_surface(struct wlr_surface *surface,
-		double lx, double ly, float rotation, struct layout_data *layout_data,
-		wlr_surface_iterator_func_t iterator, void *user_data) {
-	layout_data->x = lx;
-	layout_data->y = ly;
-	layout_data->width = surface->current.width;
-	layout_data->height = surface->current.height;
-	layout_data->rotation = rotation;
+	if (!wlr_surface_has_buffer(surface)) {
+		return false;
+	}
+
+	int sw = surface->current.width;
+	int sh = surface->current.height;
+
+	double _sx = sx + surface->sx;
+	double _sy = sy + surface->sy;
+	rotate_child_position(&_sx, &_sy, sw, sh, data->width, data->height,
+		data->rotation);
+
+	struct wlr_box box = {
+		.x = data->ox + _sx,
+		.y = data->oy + _sy,
+		.width = sw,
+		.height = sh,
+	};
+	if (surface_box != NULL) {
+		memcpy(surface_box, &box, sizeof(struct wlr_box));
```

Comment:
This could just be a normal struct assignment.

Reply:
Indeed, fixed.